experimental/transliteration/src/provider.rs

```diff
@@ -23,109 +23,163 @@ use icu_collections::{
-    #[serde(borrow)]
+    #[cfg_attr(feature = "serde", serde(borrow))]
     pub rule_group_list: VarZeroVec<'a, VarZeroSlice<RuleULE>>,
+    /// The direct dependencies of this transliterator.
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    pub dependencies: VarZeroVec<'a, str>,
 }
 
 /// The ID of a transliterator plus an optional filter.
 #[derive(Debug, Clone)]
 #[make_varule(SimpleIdULE)]
 #[zerovec::skip_derive(Ord)]
 #[zerovec::derive(Debug)]
-#[zerovec::derive(Serialize, Deserialize)]
-#[derive(serde::Serialize, serde::Deserialize)]
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Deserialize),
+    zerovec::derive(Deserialize)
+)]
+#[cfg_attr(
+    feature = "datagen",
+    derive(serde::Serialize),
+    zerovec::derive(Serialize)
+)]
 pub struct SimpleId<'a> {
     /// The filter for the transliterator. If there is none, the set of all code points is used.
     #[zerovec::varule(CodePointInversionListULE)]
-    #[serde(borrow)]
+    #[cfg_attr(feature = "serde", serde(borrow))]
     pub filter: CodePointInversionList<'a>,
     /// The ID of the transliterator.
-    #[serde(borrow)]
+    #[cfg_attr(feature = "serde", serde(borrow))]
     pub id: Cow<'a, str>,
 }
 
 /// A conversion rule. The source patterns as well as the replacer use inlined private use characters
 /// that refer to elements of the [`VarTable`] for special matchers (variables, UnicodeSets, ...).
+#[derive(Debug, Clone)]
 #[make_varule(RuleULE)]
 #[zerovec::skip_derive(Ord)]
 #[zerovec::derive(Debug)]
-#[zerovec::derive(Serialize, Deserialize)]
-#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Deserialize),
+    zerovec::derive(Deserialize)
+)]
+#[cfg_attr(
+    feature = "datagen",
+    derive(serde::Serialize),
+    zerovec::derive(Serialize)
+)]
 pub struct Rule<'a> {
     /// The pattern for the ante context. This is not replaced.
-    #[serde(borrow)]
+    #[cfg_attr(feature = "serde", serde(borrow))]
     pub ante: Cow<'a, str>,
     /// The pattern for the key. This is what gets replaced.
-    #[serde(borrow)]
+    #[cfg_attr(feature = "serde", serde(borrow))]
     pub key: Cow<'a, str>,
     /// The pattern for the post context. This is not replaced.
-    #[serde(borrow)]
+    #[cfg_attr(feature = "serde", serde(borrow))]
     pub post: Cow<'a, str>,
     /// The replacer. The key gets replaced with this.
-    #[serde(borrow)]
+    #[cfg_attr(feature = "serde", serde(borrow))]
     pub replacer: Cow<'a, str>,
     /// The offset of the cursor after this rule, if the rule matches.
     /// The end of the key/replacer is 0, i.e., a no-op.
     pub cursor_offset: i32,
 }
 
 /// The special matchers and replacers used by this transliterator.
-#[derive(Debug, Clone, serde::Serialize, serde::Deserialize, PartialEq, Eq)]
+#[derive(
+    Debug,
+    Clone,
+    zerofrom::ZeroFrom, // QUESTION: does ZeroFrom need to be behind a feature?
```

Comment:
It's required by `DataPayload::with_mut` and `DataPayload::try_map_project`, which is used by `.as_deserializing()` only, so it could be behind the `serde` feature. However, it's a dependency of `icu_provider`, so making it optional doesn't gain much, and we don't have it behind a feature anywhere.